### PR TITLE
Adding negative caching

### PR DESCRIPTION
**Pull request by cviecco**

### CODE REVIEW

File: broker/aws/impl.go [hunk 1]
```diff
@@ -387,6 +388,10 @@ func (b *Broker) getAWSRolesForAccount(accountName string) ([]string, error) {
 			b.logger.Debugf(1, "Got roles from cache")
 			return cachedEntry.Roles, nil
 		}
+		if cachedEntry.LastBadTime.After(time.Now().Add(time.Second * (-1 * negativeCacheSeconds))) {
+			b.logger.Debugf(1, "Returning recently failed from cache")
```
Comment (rgooch) [inline broker/aws/impl.go:392]: Log message doesn't read well.

File: broker/aws/impl.go [hunk 2]
```diff
@@ -387,6 +388,10 @@ func (b *Broker) getAWSRolesForAccount(accountName string) ([]string, error) {
 			b.logger.Debugf(1, "Got roles from cache")
 			return cachedEntry.Roles, nil
 		}
+		if cachedEntry.LastBadTime.After(time.Now().Add(time.Second * (-1 * negativeCacheSeconds))) {
```
Comment (rgooch) [inline broker/aws/impl.go:391]: You can just do `-negativeCacheSeconds`.

File: broker/aws/impl.go [hunk 3]
```diff
@@ -374,6 +374,7 @@ func (b *Broker) getAWSRolesForAccountNonCached(accountName string) ([]string, e
 }
 
 const roleCacheDuration = time.Second * 1800
+const negativeCacheSeconds = 60
```
Comment (rgooch) [inline broker/aws/impl.go:377]: How about 15 seconds?

File: broker/aws/api.go [hunk 4]
```diff
@@ -18,8 +18,9 @@ type userAllowedCredentialsCacheEntry struct {
 }
 
 type accountRoleCacheEntry struct {
-	Roles      []string
-	Expiration time.Time
+	Roles       []string
```
Comment (rgooch) [inline broker/aws/api.go:21]: Should these be protected by a lock?

**Review by rgooch — CHANGES_REQUESTED** [review]

File: broker/aws/api.go (hunk 4 above)
Comment (cviecco) [reply broker/aws/api.go:21]: no need the map is the one protected.

File: broker/aws/impl.go (hunk 3 above)
Comment (cviecco) [reply broker/aws/impl.go]: done

File: broker/aws/impl.go (hunk 2 above)
Comment (cviecco) [reply broker/aws/impl.go]: done

**Pull request merged**